### PR TITLE
Make test_torch.py pass cuda-memcheck

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Make the following changes:
- When there are more than 10k errors, cuda-memcheck only shows 10k errors, in this case we shouldn't raise an Exception
- Add UNDER_CUDA_MEMCHECK environment to allow disabling `pin_memory` tests when running cuda-memcheck. 
- Add a `--ci` command option, when turned on, then this script would run output to stdout instead of writing a file, and exit with an error if cuda-memcheck fails
- Add a `--nohang` command option. When turned on, then hang would be treated as pass instead of error
- Do simple filtering on the test to run: if `'cpu'` in the test name but not `'cuda'` is not in the test name
- Add `--split` and `--rank` to allowing splitting the work (NVIDIA CI has a limitation of 3 hours, we have to split the work to satisfy this limitation)
- The error summary could be `ERROR SUMMARY: 1 error`, or `ERROR SUMMARY: 2 errors`, the tail could be `error` or `errors`, it is not of the same length. The script is fixed to handle this case.
- Ignore errors from `cufft`

### CODE REVIEW

File: test/scripts/run_cuda_memcheck.py [hunk 1]
```diff
@@ -108,13 +145,15 @@ def is_ignored_only(output):
...
                     print(stderr, file=logfile)
+                    if args.ci:
+                        sys.exit("Failure detected on cuda-memcheck")
                 else:
                     print('Ignored:', test, file=logfile)
         del proc
         progressbar.update(1)
 
 async def main():
-    tasks = [asyncio.create_task(run1(i)) for i in range(args.nproc)]
```
Comment (zasdfgbnm) [inline test/scripts/run_cuda_memcheck.py]: `create_task` is available only on python 3.7...

File: test/common_utils.py [hunk 2]
```diff
@@ -280,6 +280,8 @@ def _check_module_exists(name):
 # it felt a little awkward.
 TEST_SKIP_FAST = os.getenv('PYTORCH_TEST_SKIP_FAST', '0') == '1'
 
+UNDER_CUDA_MEMCHECK = os.getenv('UNDER_CUDA_MEMCHECK', '0') == '1'
```
Comment (mruberry) [inline test/common_utils.py]: Can you put this in common_device_type near the CUDA class? 

Also, can we name it just CUDA_MEMCHECK?

Comment (ngimel) [reply test/common_utils.py]: Actually, can we name the env var PYTORCH_CUDA_MEMCHECK, to avoid possible collisions?

Comment (zasdfgbnm) [reply test/common_utils.py]: fixed

**Comment by zasdfgbnm** [comment]
I forgot to `import os`...

**Review by mruberry — APPROVED** [review]
Awesome! Thanks!

**Pull request closed without merging**